### PR TITLE
test: Flutter test harness + critical notifier/filter unit tests

### PR DESCRIPTION
**Pull request by AndreaDiazCorreia**
Refs #152 (PR1 of 2 — phased).

First slice of the Flutter test base:

- `ProviderContainer` harness with auto-dispose + fake data builders (`test/support/`).
- Notifier tests: `BackupReminderNotifier`/`BackupCompletedNotifier` (snooze, dismiss, legacy migration) and `FirstRunNotifier` (loading→data, first-launch default).
- Filter/derivation tests: `filteredOrdersProvider` (tabs, ownership, currency/payment/rating/premium, sort) and `filteredTradesWithOrderStateProvider` (status filter, terminal-status collapse, sort) — Rust-backed providers are overridden with fakes.

29 tests passing; `flutter analyze` clean. Suite runs in the existing CI `flutter test` step (no workflow change needed).

Deferred to PR2: DESIGN_SYSTEM golden tests.

## Summary by CodeRabbit

- **Tests**
  - Added coverage for backup reminders and completion, including load/reconcile behavior, snoozing, dismissal, persistence, and legacy fallback.
  - Added first-run completion notifier tests covering loading, defaults, and persistence.
  - Added Riverpod provider tests for orders and trades filtering, status mapping, sorting, and edge cases (unrated/range orders, terminal status bucketing).
- **Test Support**
  - Added deterministic order/trade fixtures, including “fake now” timestamping and range order support.
  - Introduced Riverpod test utilities (container creation and order book harness) to streamline provider-based assertions.

### CODE REVIEW

**Comment by AndreaDiazCorreia** [comment]
@codex review

**Comment by grunch** [comment]
@coderabbitai review

**Review by grunch — COMMENTED** [review]
## Strict Review

**Actionable comments: 9** (1 high, 4 medium, 4 low) · **Verdict: Approve with observations** — no blocking issues; the HIGH is a one-line matcher fix worth doing before merge.

### Verification performed
- ✅ Every assertion cross-checked against the actual provider sources (`backup_reminder_provider.dart`, `home_order_providers.dart`, `trades_providers.dart`, `first_run_provider.dart`) — they reflect implemented behavior (prefs keys, snooze semantics, terminal-status collapse, default-range guards).
- ✅ `flutter analyze` run locally on this branch in an isolated worktree: **clean**.
- ✅ CI green (Flutter analyze/test, Rust, CodeRabbit); the 29-test count in the PR body checks out (12+9+4+4).

### What's solid
- `createContainer` + `addTearDown(container.dispose)` is the canonical Riverpod test harness; zero state leakage between tests (`setMockInitialValues` resets prefs per test).
- Rust-backed providers correctly isolated via `overrideWith` — no native bridge involved.
- `kFakeNow` anchored timestamps make sort assertions deterministic.
- Behavior-descriptive test names, consistent AAA structure, legacy fallback covered (`BackupCompletedNotifier` → dismissed flag), and the non-obvious `initialValue` + live-snooze reconciliation case is tested.
- Fits existing CI with no workflow change; accurate PR description.

### Summary of findings (details inline)
| Severity | File | Issue |
|---|---|---|
| 🔴 High | `filtered_trades_provider_test.dart` | Terminal-collapse test uses `containsAll` — cannot fail for the exact regression it targets |
| 🟠 Medium | `filtered_trades_provider_test.dart` | `"All"` test: `containsAll` doesn't assert length |
| 🟠 Medium | `filtered_trades_provider_test.dart` | `orderStatusToFilter` has 15 arms; only the success bucket is partially covered |
| 🟠 Medium | `filtered_orders_provider_test.dart` | Default-range guard semantics (`!= defaultRatingRange`) unpinned |
| 🟠 Medium | `fake_orders.dart` | Builder cannot produce range-shaped orders |
| 🟡 Low | `filtered_orders_provider_test.dart` | Latent `autoDispose` fragility in `_bookWith` |
| 🟡 Low | `filtered_orders_provider_test.dart` | `ProviderContainerHelper` could live in `test/support/` |
| 🟡 Low | `fake_trades.dart` | Fixture pairs `status: active` with `TradeStep.disputed()` |
| 🟡 Low | `backup_reminder_provider_test.dart` | Wall-clock dependence in snooze fixtures (noted, not blocking) |

File: test/features/trades/filtered_trades_provider_test.dart [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        () async {
+      final container = _tradesWith([
+        fakeTrade(id: 'success', status: OrderStatus.success),
+        fakeTrade(id: 'settled', status: OrderStatus.settledByAdmin),
+        fakeTrade(id: 'canceled', status: OrderStatus.canceled),
+      ]);
+
+      expect(
+        await _orderIds(container, filter: TradeStatusFilter.success),
+        containsAll(['order-success', 'order-settled']),
```
Comment (grunch) [inline test/features/trades/filtered_trades_provider_test.dart]: _⚠️ Potential issue_ | _🔴 High_

**This test cannot fail for the exact regression it targets.**

The fixture deliberately includes a `canceled` trade to prove it is **excluded** from the `success` filter, but `containsAll` passes even if `order-canceled` leaks through. If `orderStatusToFilter` regressed and mapped `canceled → success`, this test would stay green. Use `unorderedEquals` to assert the exact result set:

```suggestion
        unorderedEquals(['order-success', 'order-settled']),
```

File: test/features/trades/filtered_trades_provider_test.dart [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+void main() {
+  group('filteredTradesWithOrderStateProvider', () {
+    test('"All" returns every trade', () async {
+      final container = _tradesWith([
+        fakeTrade(id: 'a', status: OrderStatus.active),
+        fakeTrade(id: 'b', status: OrderStatus.pending),
+      ]);
+
+      expect(await _orderIds(container), containsAll(['order-a', 'order-b']));
```
Comment (grunch) [inline test/features/trades/filtered_trades_provider_test.dart]: _⚠️ Potential issue_ | _🟠 Medium_

**Same weak-matcher pattern: `containsAll` doesn't assert length.**

Duplicated or extra synthesized items would pass unnoticed. Prefer the exact set:

```suggestion
      expect(await _orderIds(container), unorderedEquals(['order-a', 'order-b']));
```

File: test/features/trades/filtered_trades_provider_test.dart [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  ProviderContainer container, {
+  TradeStatusFilter filter = TradeStatusFilter.all,
+}) async {
+  container.read(selectedStatusFilterProvider.notifier).state = filter;
+  final items = await container.read(filteredTradesWithOrderStateProvider.future);
+  return items.map((t) => t.orderId).toList();
+}
+
+void main() {
+  group('filteredTradesWithOrderStateProvider', () {
```
Comment (grunch) [inline test/features/trades/filtered_trades_provider_test.dart:25]: _🧪 Test coverage_ | _🟠 Medium_

**`orderStatusToFilter` has 15 arms; only the success bucket is (partially) covered.**

Untested: the `canceled` bucket (`canceled`/`expired`/`cooperativelyCanceled`/`canceledByAdmin`), `dispute`, `settledHoldInvoice`, `completedByAdmin`, and `inProgress → active`. This is pure protocol-mapping logic — an exhaustive table test is ~20 lines and pins the whole mapping:

```dart
test('orderStatusToFilter maps every protocol status to its bucket', () {
  const expected = {
    OrderStatus.pending: TradeStatusFilter.pending,
    OrderStatus.waitingBuyerInvoice: TradeStatusFilter.waitingInvoice,
    OrderStatus.waitingPayment: TradeStatusFilter.waitingPayment,
    OrderStatus.active: TradeStatusFilter.active,
    OrderStatus.inProgress: TradeStatusFilter.active,
    OrderStatus.fiatSent: TradeStatusFilter.fiatSent,
    OrderStatus.settledHoldInvoice: TradeStatusFilter.success,
    OrderStatus.success: TradeStatusFilter.success,
    OrderStatus.settledByAdmin: TradeStatusFilter.success,
    OrderStatus.completedByAdmin: TradeStatusFilter.success,
    OrderStatus.canceled: TradeStatusFilter.canceled,
    OrderStatus.expired: TradeStatusFilter.canceled,
    OrderStatus.cooperativelyCanceled: TradeStatusFilter.canceled,
    OrderStatus.canceledByAdmin: TradeStatusFilter.canceled,
    OrderStatus.dispute: TradeStatusFilter.dispute,
  };
  expected.forEach((status, bucket) {
    expect(orderStatusToFilter(status), bucket, reason: '$status');
  });
});
```

Reasonable to include in this PR since it is *the* test-foundation PR.

File: test/features/home/filtered_orders_provider_test.dart [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        fakeOrder(id: 'cash', kind: 'sell', paymentMethod: 'Cash'),
+      ]);
+      helper.setTab(OrderType.buy);
+      helper.container.read(paymentMethodFilterProvider.notifier).state =
+          ['revolut'];
+
+      expect(helper.ids(), ['multi']);
+    });
+
+    test('rating range excludes orders outside the bounds', () async {
```
Comment (grunch) [inline test/features/home/filtered_orders_provider_test.dart:72]: _🧪 Test coverage_ | _🟠 Medium_

**The default-range guard semantics are unpinned.**

In `filteredOrdersProvider`, when the rating/premium range equals its default (`!= defaultRatingRange` guard), the filter is not applied at all — an order with `rating: 0.0` still shows. That's subtle, intentional behavior no test protects: if someone "simplifies" the guard away, nothing fails. A cheap companion test pins it:

```dart
test('default rating range shows unrated orders', () async {
  final helper = await _bookWith([
    fakeOrder(id: 'unrated', kind: 'sell', rating: 0.0),
  ]);
  helper.setTab(OrderType.buy);

  expect(helper.ids(), ['unrated']); // guard: default range = no filtering
});
```

File: test/features/home/filtered_orders_provider_test.dart [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+import '../../support/fake_orders.dart';
+import '../../support/provider_harness.dart';
+
+/// Awaits the book so [filteredOrdersProvider] can be read synchronously after.
+Future<ProviderContainerHelper> _bookWith(List<OrderItem> orders) async {
+  final container = createContainer(overrides: [
+    orderBookProvider.overrideWith((ref) => Stream.value(orders)),
+  ]);
+  await container.read(orderBookProvider.future);
```
Comment (grunch) [inline test/features/home/filtered_orders_provider_test.dart]: _🛠️ Refactor suggestion_ | _🟡 Low_

**Latent `autoDispose` fragility.**

`orderBookProvider` is `autoDispose`, and between this `await` and the first `ids()` read there is no live listener. Today it works because `setTab`/`ids()` run synchronously after the await — but if a future test inserts another `await` in between, the provider can be disposed and rebuilt in loading state → empty list → confusing failure. Keeping a listener alive makes the harness robust:

```suggestion
  // Keep the autoDispose stream provider alive across later awaits.
  container.listen(orderBookProvider, (_, __) {});
  await container.read(orderBookProvider.future);
```

File: test/features/home/filtered_orders_provider_test.dart [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        fakeOrder(id: 'newer', kind: 'sell', minutesAgo: 5),
+      ]);
+      helper.setTab(OrderType.buy);
+
+      expect(helper.ids(), ['newer', 'older']);
+    });
+  });
+}
+
+class ProviderContainerHelper {
```
Comment (grunch) [inline test/features/home/filtered_orders_provider_test.dart]: _🧹 Nitpick_ | _🟡 Low_

**Consider moving `ProviderContainerHelper` to `test/support/`.**

If the trades tests (or PR2's golden tests) grow a similar helper, this is the natural shared home — same rationale as `provider_harness.dart`. Fine to defer.

File: test/support/fake_orders.dart [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+/// Anchor so timestamp-based sort assertions don't depend on the wall clock.
+final DateTime kFakeNow = DateTime.utc(2026, 1, 1, 12);
+
+/// Builds a fixed-amount [OrderItem], exposing only the fields filter logic
+/// reads. [minutesAgo] controls recency for sort assertions.
+OrderItem fakeOrder({
+  String id = 'order-1',
+  String kind = 'sell',
+  double fiatAmount = 100,
```
Comment (grunch) [inline test/support/fake_orders.dart]: _🧪 Test coverage_ | _🟠 Medium_

**`fakeOrder` cannot build range-shaped orders.**

With `fiatAmount` as a non-nullable `double = 100`, the range shape (`fiatAmountMin`/`fiatAmountMax`), the `ArgumentError` shape validation in the `OrderItem` constructor, and `displayAmount` for ranges are all untestable through this builder. Acceptable for PR1, but worth extending before more tests accumulate on top of it, e.g.:

```dart
OrderItem fakeOrder({
  // ...
  double? fiatAmount = 100,
  double? fiatAmountMin,
  double? fiatAmountMax,
  // ...
}) {
  final isRange = fiatAmountMin != null && fiatAmountMax != null;
  return OrderItem(
    // ...
    fiatAmount: isRange ? null : fiatAmount,
    fiatAmountMin: fiatAmountMin,
    fiatAmountMax: fiatAmountMax,
    // ...
  );
}
```

File: test/support/fake_trades.dart [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    createdAt: startedAt,
+    isMine: isMine,
+  );
+
+  return TradeInfo(
+    id: id,
+    order: order,
+    role: role,
+    counterpartyPubkey: 'counterparty-$id',
+    currentStep: const TradeStep.disputed(),
```
Comment (grunch) [inline test/support/fake_trades.dart:33]: _🧹 Nitpick_ | _🟡 Low_

**Internally inconsistent fixture: `status: active` paired with `TradeStep.disputed()`.**

Verified that the trades-list mapping never reads `currentStep` today, so this is harmless — but an "arbitrary" value that contradicts the trade's status is a trap for future tests reusing this builder. Prefer a neutral step or make it a parameter:

```dart
TradeInfo fakeTrade({
  // ...
  TradeStep currentStep = const TradeStep.buyer(BuyerStep.waitingForSellerPayment),
}) {
```

(or whichever `BuyerStep` variant is the natural neutral value).

File: test/features/account/backup_reminder_provider_test.dart [hunk 9]
```diff
@@ -0,0 +1,174 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mostro/features/account/providers/backup_reminder_provider.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+Future<SharedPreferences> _prefs() => SharedPreferences.getInstance();
+
+int _inFuture(Duration d) => DateTime.now().add(d).millisecondsSinceEpoch;
+int _inPast(Duration d) => DateTime.now().subtract(d).millisecondsSinceEpoch;
```
Comment (grunch) [inline test/features/account/backup_reminder_provider_test.dart:8]: _🧹 Nitpick_ | _🟡 Low_

**Wall-clock dependence — noted, not blocking.**

`_inFuture`/`_inPast` compute against `DateTime.now()` at setup while production compares against `now()` inside `load()`. With 12h/1h margins the flake risk is negligible; flagging it only because the codebase has no clock abstraction yet, so this pattern will spread. If snooze logic ever gets sub-minute windows, inject a clock instead.

**Review by grunch — APPROVED** [review]
All High/Medium comments resolved and verified against the actual code:

- 🔴 High (success test did not verify exclusion) → `unorderedEquals` + the fixture keeps the `canceled` trade, so it now fails if one leaks into the success bucket.
- 🟠 `containsAll`→`unorderedEquals`, exhaustive `orderStatusToFilter` test with a length guard, coverage for unrated/range orders, and `fakeOrder` now supports range orders.
- 🟡 autoDispose keepAlive + helper extracted to `test/support/order_book_harness.dart`.

CI green (Flutter, Rust, CodeRabbit). The two remaining Low nits (active+disputed fixture in fake_trades, wall-clock in backup_reminder) are optional follow-ups. LGTM ✅
